### PR TITLE
feat: allow hooks for global and bulk smart action

### PR DESCRIPTION
**Pull request by arnaud-moncel**
## Pull Request checklist:

- [X] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [ ] Create automatic tests
- [X] Test manually the implemented changes
- [X] Review my own code (indentation, syntax, style, simplicity, readability)
- [ ] Wonder if you can improve the existing code
- [X] If needed, Types have been updated

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -32,9 +32,31 @@ export interface DatabaseConfiguration {
...
+      ids: string[],
+      parent_collection_name: string,
+      parent_collection_id: string,
+      parent_association_name: string,
+      all_records: boolean,
+      all_records_subset_query: Query,
+      all_records_ids_excluded: string[],
+      smart_action_id: string,
+    },
+    type: string,
```
Comment (SteveBunlon) [inline types/index.d.ts:52]:
```suggestion
    type: 'custom-action-requests',
```

**Review by SteveBunlon — APPROVED** [review]
Nice thanks for the update 😄 

Just one thing to fix and you can 🚀

File: types/index.d.ts [hunk 2]
```diff
@@ -222,16 +244,20 @@ export interface SmartActionLoadHookField extends SmartActionHookField {
   position: number,
 }
 
+export interface SmartActionHookRequest extends ForestRequest {
+  body: SelectAllRequestBody,
+}
+
 export interface SmartActionLoadHook<T = any> {
-  (context: { fields: SmartActionLoadHookField[], record: T & Document }): SmartActionLoadHookField[]
+  (context: { fields: SmartActionLoadHookField[], request: SmartActionHookRequest }): SmartActionLoadHookField[]
```
Comment (SteveBunlon) [inline types/index.d.ts:252]: Actually, the T generic type is not used anymore. If we don't provide any record to the user, you can remove it.

File: types/index.d.ts [hunk 3]
```diff
@@ -222,16 +244,20 @@ export interface SmartActionLoadHookField extends SmartActionHookField {
...
 }
 
 export interface SmartActionChangeHookField extends SmartActionHookField {
   previousValue: any,
 }
 
 export interface SmartActionChangeHook<T = any> {
-  (context: { fields: SmartActionChangeHookField[], record: T, changedField: SmartActionChangeHookField }): SmartActionChangeHookField[]
+  (context: { fields: SmartActionChangeHookField[], changedField: SmartActionChangeHookField, request: SmartActionHookRequest }): SmartActionChangeHookField[]
```
Comment (SteveBunlon) [inline types/index.d.ts:260]: Actually, the T generic type is not used anymore. If we don't provide any record to the user, you can remove it.

**Review by SteveBunlon — CHANGES_REQUESTED** [review]

File: types/index.d.ts (hunk 1 above)
Comment (arnaud-moncel) [reply types/index.d.ts:52]: I don't know it is possible 😱  thanks 😀

File: types/index.d.ts [hunk 4]
```diff
@@ -32,9 +32,31 @@ export interface DatabaseConfiguration {
   }
 }
 
+export interface ForestRequest extends Request {
+  user: User,
+}
+
+export interface SelectAllRequestBody {
+  data: {
```
Comment (SteveBunlon) [inline types/index.d.ts:40]: The `values` attribute is not present anymore ?

File: types/index.d.ts [hunk 5]
```diff
@@ -63,7 +85,7 @@ export class RecordGetter<T> extends AbstractRecordTool<T> {
 
 export class RecordsGetter<T> extends AbstractRecordTool<T> {
   getAll(query: Query): Promise<(T & Document)[]>;
-  getIdsFromRequest(request: Request): Promise<string[]>;
+  getIdsFromRequest(request: ForestRequest): Promise<string[]>;
```
Comment (SteveBunlon) [inline types/index.d.ts:88]: [THINKING] Couldn't it be the same as SmartActionHookRequest ?

**Review by SteveBunlon — CHANGES_REQUESTED** [review]

**Review by SteveBunlon — DISMISSED** [review]
🔥 🔥 🔥 🔥 🔥 🔥 🔥

**Review by SteveBunlon — APPROVED** [review]
🏌️

**Pull request merged**